Library/Homebrew/dependencies_helpers.rb

```diff
@@ -34,10 +34,20 @@ def args_includes_ignores(args)
     [includes, ignores]
   end
 
-  def recursive_includes(klass, root_dependent, includes, ignores)
+  def recursive_includes(
+    klass,
+    root_dependent,
+    includes,
+    ignores,
+    used_formulae = [],
+    skip_recursive_build_dependents: false
+  )
     raise ArgumentError, "Invalid class argument: #{klass}" if klass != Dependency && klass != Requirement
 
-    cache_key = "recursive_includes_#{includes}_#{ignores}"
+    cache_key = +"recursive_includes_#{includes.join("_")}_#{ignores.join("_")}"
+    cache_key << "_#{root_dependent}" if includes.include?("test?")
+    cache_key << "_#{root_dependent}_no_recursive_build_#{used_formulae.join("_")}" if skip_recursive_build_dependents
+    cache_key.freeze
```

Comment:
The `join`s are technically not needed, but they help a lot with readability of the `cache_key` (which I did a little of while trying to debug some discrepancies with caching and without).